diff -up ./data/firewalld/meson.build.orig ./data/firewalld/meson.build
--- ./data/firewalld/meson.build.orig	2020-09-19 23:13:39.000000000 +0200
+++ ./data/firewalld/meson.build	2020-09-20 14:41:43.541263701 +0200
@@ -2,7 +2,7 @@
 if get_option('firewalld')
   install_data(
     'gsconnect.xml',
-    install_dir: join_paths(libdir, 'firewalld', 'services')
+    install_dir: join_paths(prefix, 'lib', 'firewalld', 'services')
   )
 endif
 
