diff --git a/extension.js b/extension.js
index a371565..240edaa 100644
--- a/extension.js
+++ b/extension.js
@@ -50,6 +50,7 @@ class LogoMenuMenuButton extends PanelMenu.Button {
         // Menu
         this._settings.connectObject('changed::hide-softwarecentre', () => this._displayMenuItems(), this);
         this._settings.connectObject('changed::show-power-options', () => this._displayMenuItems(), this);
+        this._settings.connectObject('changed::show-boxbuddy', () => this._displayMenuItems(), this);
         this._settings.connectObject('changed::hide-forcequit', () => this._displayMenuItems(), this);
         this._settings.connectObject('changed::show-lockscreen', () => this._displayMenuItems(), this);
         this._settings.connectObject('changed::show-activities-button', () => this._displayMenuItems(), this);
@@ -65,6 +66,7 @@ class LogoMenuMenuButton extends PanelMenu.Button {
 
     _displayMenuItems() {
         const showPowerOptions = this._settings.get_boolean('show-power-options');
+        const showBoxBuddy = this._settings.get_boolean('show-boxbuddy');
         const showForceQuit = !this._settings.get_boolean('hide-forcequit');
         const showLockScreen = this._settings.get_boolean('show-lockscreen');
         const showSoftwareCenter = !this._settings.get_boolean('hide-softwarecentre');
@@ -73,21 +75,26 @@ class LogoMenuMenuButton extends PanelMenu.Button {
         this.menu.removeAll();
 
         this._addItem(new MenuItem(_('About My System'), () => this._aboutThisDistro()));
-        // this._addItem(new MenuItem(_('System Settings...'), () => this._systemPreferences()));
+        this._addItem(new MenuItem(_('Documentation'), () => this._documentation()));
         this._addItem(new PopupMenu.PopupSeparatorMenuItem());
 
+        this._addItem(new MenuItem(_('System Settings...'), () => this._systemPreferences()));
+
         if (!showActivitiesButton)
             this._addItem(new MenuItem(_('Activities'), () => this._overviewToggle()));
 
-        this._addItem(new MenuItem(_('App Grid'), () => this._showAppGrid()));
-        this._addItem(new PopupMenu.PopupSeparatorMenuItem());
-
+        // this._addItem(new MenuItem(_('App Grid'), () => this._showAppGrid()));
         if (showSoftwareCenter)
             this._addItem(new MenuItem(_('Software Center'), () => this._openSoftwareCenter()));
+        this._addItem(new PopupMenu.PopupSeparatorMenuItem());
 
+        this._addItem(new MenuItem(_('Extensions'), () => this._openExtensionsApp()));
         this._addItem(new MenuItem(_('System Monitor'), () => this._openSystemMonitor()));
         this._addItem(new MenuItem(_('Terminal'), () => this._openTerminal()));
-        this._addItem(new MenuItem(_('Extensions'), () => this._openExtensionsApp()));
+        this._addItem(new PopupMenu.PopupSeparatorMenuItem());
+
+        if (showBoxBuddy)
+            this._addItem(new MenuItem(_('Containers'), () => this._openBoxBuddy()));
 
         if (showForceQuit) {
             this._addItem(new PopupMenu.PopupSeparatorMenuItem());
@@ -133,6 +140,10 @@ class LogoMenuMenuButton extends PanelMenu.Button {
         Util.spawn(['gnome-control-center']);
     }
 
+    _documentation() {
+        Util.trySpawnCommandLine('xdg-open https://docs.projectbluefin.io/')
+    }
+
     _overviewToggle() {
         Main.overview.toggle();
     }
@@ -175,6 +186,10 @@ class LogoMenuMenuButton extends PanelMenu.Button {
         Util.trySpawnCommandLine(this._settings.get_string('menu-button-terminal'));
     }
 
+    _openBoxBuddy() {
+        Util.trySpawnCommandLine('flatpak run io.github.dvlv.boxbuddyrs');
+    }
+
     _openSoftwareCenter() {
         Util.trySpawnCommandLine(this._settings.get_string('menu-button-software-center'));
     }

