diff --git a/constants.js b/constants.js
index fcd799c..b30ca24 100644
--- a/constants.js
+++ b/constants.js
@@ -31,6 +31,8 @@ export const SymbolicDistroIcons = [
     {PATH: '/Resources/steam-deck-logo-symbolic.svg'},
     {PATH: '/Resources/ublue-logo-symbolic.svg'},
     {PATH: '/Resources/centos-logo-symbolic.svg'},
+    {PATH: '/Resources/framework-logo-symbolic.svg'},
+    {PATH: '/Resources/ampere-logo-symbolic.svg'},
 ];
 
 export const ColouredDistroIcons = [
