diff --git a/src/vte.cc b/src/vte.cc
index 82137b96..6e543a3b 100644
--- a/src/vte.cc
+++ b/src/vte.cc
@@ -10240,7 +10240,7 @@ Terminal::emit_pending_signals()
 
 	emit_adjustment_changed();
 
-#if _VTE_GTK == 3
+//#if _VTE_GTK == 3
         if (m_pending_changes & vte::to_integral(PendingChanges::NOTIFICATION)) {
                 _vte_debug_print (VTE_DEBUG_SIGNALS,
                                   "Emitting `notification-received'.\n");
@@ -10268,7 +10268,7 @@ Terminal::emit_pending_signals()
                 g_object_notify_by_pspec(freezer.get(), pspecs[PROP_CURRENT_CONTAINER_NAME]);
                 g_object_notify_by_pspec(freezer.get(), pspecs[PROP_CURRENT_CONTAINER_RUNTIME]);
         }
-#endif
+//#endif
 
 	if (m_pending_changes & vte::to_integral(PendingChanges::TITLE)) {
                 if (m_window_title != m_window_title_pending) {
diff --git a/src/vte/vteterminal.h b/src/vte/vteterminal.h
index 5afb3b24..ce2d0559 100644
--- a/src/vte/vteterminal.h
+++ b/src/vte/vteterminal.h
@@ -109,11 +109,11 @@ struct _VteTerminalClass {
 
 	void (*bell)(VteTerminal* terminal);
 
-#if _VTE_GTK == 3
+//#if _VTE_GTK == 3
 	void (*notification_received)(VteTerminal* terminal, const gchar *summary, const gchar *body);
 	void (*shell_precmd)(VteTerminal* terminal);
 	void (*shell_preexec)(VteTerminal* terminal);
-#endif /* _VTE_GTK == 3 */
+//#endif /* _VTE_GTK == 3 */
 
         /* Add new vfuncs here, and subtract from the padding below. */
 
@@ -573,12 +573,12 @@ _VTE_PUBLIC
 glong vte_terminal_get_column_count(VteTerminal *terminal) _VTE_CXX_NOEXCEPT _VTE_GNUC_NONNULL(1);
 _VTE_PUBLIC
 const char *vte_terminal_get_window_title(VteTerminal *terminal) _VTE_CXX_NOEXCEPT _VTE_GNUC_NONNULL(1);
-#if _VTE_GTK == 3
+//#if _VTE_GTK == 3
 _VTE_PUBLIC
 const char *vte_terminal_get_current_container_name(VteTerminal *terminal) _VTE_CXX_NOEXCEPT _VTE_GNUC_NONNULL(1);
 _VTE_PUBLIC
 const char *vte_terminal_get_current_container_runtime(VteTerminal *terminal) _VTE_CXX_NOEXCEPT _VTE_GNUC_NONNULL(1);
-#endif
+//#endif
 _VTE_PUBLIC
 const char *vte_terminal_get_current_directory_uri(VteTerminal *terminal) _VTE_CXX_NOEXCEPT _VTE_GNUC_NONNULL(1);
 _VTE_PUBLIC
diff --git a/src/vtegtk.cc b/src/vtegtk.cc
index 218b5b08..7ae5bf7e 100644
--- a/src/vtegtk.cc
+++ b/src/vtegtk.cc
@@ -963,14 +963,14 @@ try
                 case PROP_CURSOR_BLINK_MODE:
                         g_value_set_enum (value, vte_terminal_get_cursor_blink_mode (terminal));
                         break;
-#if _VTE_GTK == 3
+//#if _VTE_GTK == 3
                 case PROP_CURRENT_CONTAINER_NAME:
                         g_value_set_string (value, vte_terminal_get_current_container_name (terminal));
                         break;
                 case PROP_CURRENT_CONTAINER_RUNTIME:
                         g_value_set_string (value, vte_terminal_get_current_container_runtime (terminal));
                         break;
-#endif
+//#endif
                 case PROP_CURRENT_DIRECTORY_URI:
                         g_value_set_string (value, vte_terminal_get_current_directory_uri (terminal));
                         break;
@@ -1288,11 +1288,11 @@ vte_terminal_class_init(VteTerminalClass *klass)
 	klass->child_exited = NULL;
 	klass->encoding_changed = NULL;
 	klass->char_size_changed = NULL;
-#if _VTE_GTK == 3
+//#if _VTE_GTK == 3
 	klass->notification_received = NULL;
 	klass->shell_precmd = NULL;
 	klass->shell_preexec = NULL;
-#endif
+//#endif
 	klass->window_title_changed = NULL;
 	klass->icon_title_changed = NULL;
 	klass->selection_changed = NULL;
@@ -1376,7 +1376,7 @@ vte_terminal_class_init(VteTerminalClass *klass)
                                    G_OBJECT_CLASS_TYPE(klass),
                                    g_cclosure_marshal_VOID__INTv);
 
-#if _VTE_GTK == 3
+//#if _VTE_GTK == 3
         /**
          * VteTerminal::notification-received:
          * @vteterminal: the object which received the signal
@@ -1430,7 +1430,7 @@ vte_terminal_class_init(VteTerminalClass *klass)
                              NULL,
                              g_cclosure_marshal_VOID__VOID,
                              G_TYPE_NONE, 0);
-#endif
+//#endif
 
         /**
          * VteTerminal::window-title-changed:
@@ -2422,7 +2422,7 @@ vte_terminal_class_init(VteTerminalClass *klass)
                                      NULL,
                                      (GParamFlags) (G_PARAM_READABLE | G_PARAM_STATIC_STRINGS | G_PARAM_EXPLICIT_NOTIFY));
 
-#if _VTE_GTK == 3
+//#if _VTE_GTK == 3
         /**
          * VteTerminal:current-container-name:
          *
@@ -2443,7 +2443,7 @@ vte_terminal_class_init(VteTerminalClass *klass)
                 g_param_spec_string ("current-container-runtime", NULL, NULL,
                                      NULL,
                                      (GParamFlags) (G_PARAM_READABLE | G_PARAM_STATIC_STRINGS | G_PARAM_EXPLICIT_NOTIFY));
-#endif
+//#endif
 
         /**
          * VteTerminal:current-directory-uri:
@@ -5253,7 +5253,7 @@ catch (...)
         return -1;
 }
 
-#if _VTE_GTK == 3
+//#if _VTE_GTK == 3
 /**
  * vte_terminal_get_current_container_name:
  * @terminal: a #VteTerminal
@@ -5303,7 +5303,7 @@ catch (...)
         vte::log_exception();
         return NULL;
 }
-#endif
+//#endif
 
 /**
  * vte_terminal_get_current_directory_uri:
diff --git a/src/vtegtk.hh b/src/vtegtk.hh
index f7a2373e..85c2d380 100644
--- a/src/vtegtk.hh
+++ b/src/vtegtk.hh
@@ -52,11 +52,11 @@ enum {
         SIGNAL_RESIZE_WINDOW,
         SIGNAL_RESTORE_WINDOW,
         SIGNAL_SELECTION_CHANGED,
-#if _VTE_GTK == 3
+//#if _VTE_GTK == 3
         SIGNAL_SHELL_PRECMD,
         SIGNAL_SHELL_PREEXEC,
         SIGNAL_NOTIFICATION_RECEIVED,
-#endif
+//#endif
         SIGNAL_WINDOW_TITLE_CHANGED,
         LAST_SIGNAL
 };
@@ -74,10 +74,10 @@ enum {
         PROP_CJK_AMBIGUOUS_WIDTH,
         PROP_CURSOR_BLINK_MODE,
         PROP_CURSOR_SHAPE,
-#if _VTE_GTK == 3
+//#if _VTE_GTK == 3
         PROP_CURRENT_CONTAINER_NAME,
         PROP_CURRENT_CONTAINER_RUNTIME,
-#endif
+//#endif
         PROP_CURRENT_DIRECTORY_URI,
         PROP_CURRENT_FILE_URI,
         PROP_DELETE_BINDING,
diff --git a/src/vteinternal.hh b/src/vteinternal.hh
index e7615246..c382d18f 100644
--- a/src/vteinternal.hh
+++ b/src/vteinternal.hh
@@ -112,7 +112,7 @@ typedef enum _VteCharacterReplacement {
         VTE_CHARACTER_REPLACEMENT_LINE_DRAWING
 } VteCharacterReplacement;
 
-#if _VTE_GTK == 3
+//#if _VTE_GTK == 3
 struct VteContainer {
 public:
         VteContainer(const std::string &name, const std::string &runtime) :
@@ -124,7 +124,7 @@ public:
         std::string m_name;
         std::string m_runtime;
 };
-#endif
+//#endif
 
 typedef struct _VtePaletteColor {
 	struct {
@@ -677,13 +677,13 @@ public:
         gboolean m_cursor_moved_pending;
         gboolean m_contents_changed_pending;
 
-#if _VTE_GTK == 3
+//#if _VTE_GTK == 3
         std::stack<VteContainer> m_containers;
 
         /* desktop notification */
         std::string m_notification_summary;
         std::string m_notification_body;
-#endif
+//#endif
 
         std::string m_window_title{};
         std::string m_current_directory_uri{};
@@ -698,12 +698,12 @@ public:
                 TITLE = 1u << 0,
                 CWD   = 1u << 1,
                 CWF   = 1u << 2,
-#if _VTE_GTK == 3
+//#if _VTE_GTK == 3
                 NOTIFICATION = 1u << 3,
                 SHELL_PREEXEC = 1u << 4,
                 SHELL_PRECMD = 1u << 5,
                 CONTAINERS = 1u << 6,
-#endif
+//#endif
         };
         unsigned m_pending_changes{0};
 
@@ -1541,11 +1541,11 @@ public:
                              int osc) noexcept;
 
         /* OSC handlers */
-#if _VTE_GTK == 3
+//#if _VTE_GTK == 3
         void handle_urxvt_extension(vte::parser::Sequence const& seq,
                                     vte::parser::StringTokeniser::const_iterator& token,
                                     vte::parser::StringTokeniser::const_iterator const& endtoken) noexcept;
-#endif
+//#endif
         void set_color(vte::parser::Sequence const& seq,
                        vte::parser::StringTokeniser::const_iterator& token,
                        vte::parser::StringTokeniser::const_iterator const& endtoken,
diff --git a/src/vteseq.cc b/src/vteseq.cc
index 666b8591..5fa72a2a 100644
--- a/src/vteseq.cc
+++ b/src/vteseq.cc
@@ -40,11 +40,11 @@
 #define ST_C0 _VTE_CAP_ST
 
 #include <algorithm>
-#if _VTE_GTK == 3
+//#if _VTE_GTK == 3
 #include <string>
 #include <unistd.h>
 #include <sys/types.h>
-#endif
+//#endif
 
 using namespace std::literals;
 
@@ -1382,7 +1382,7 @@ Terminal::delete_lines(vte::grid::row_t param)
         m_text_deleted_flag = TRUE;
 }
 
-#if _VTE_GTK == 3
+//#if _VTE_GTK == 3
 void
 Terminal::handle_urxvt_extension(vte::parser::Sequence const& seq,
                                  vte::parser::StringTokeniser::const_iterator& token,
@@ -1391,7 +1391,7 @@ Terminal::handle_urxvt_extension(vte::parser::Sequence const& seq,
         if (token == endtoken)
                 return;
 
-#if _VTE_GTK == 3
+//#if _VTE_GTK == 3
         if (*token == "container") {
                 ++token;
 
@@ -1473,7 +1473,7 @@ Terminal::handle_urxvt_extension(vte::parser::Sequence const& seq,
 
                 return;
         }
-#endif
+//#endif
 
         if (*token == "notify") {
                 ++token;
@@ -1499,7 +1499,7 @@ Terminal::handle_urxvt_extension(vte::parser::Sequence const& seq,
                 m_pending_changes |= vte::to_integral(PendingChanges::SHELL_PREEXEC);
         }
 }
-#endif
+//#endif
 
 bool
 Terminal::get_osc_color_index(int osc,
@@ -6667,11 +6667,11 @@ Terminal::OSC(vte::parser::Sequence const& seq)
                 reset_color(VTE_HIGHLIGHT_FG, VTE_COLOR_SOURCE_ESCAPE);
                 break;
 
-#if _VTE_GTK == 3
+//#if _VTE_GTK == 3
         case VTE_OSC_URXVT_EXTENSION:
                 handle_urxvt_extension(seq, it, cend);
                 break;
-#endif
+//#endif
 
         case VTE_OSC_XTERM_SET_ICON_TITLE:
         case VTE_OSC_XTERM_SET_XPROPERTY:
@@ -6714,9 +6714,9 @@ Terminal::OSC(vte::parser::Sequence const& seq)
         case VTE_OSC_URXVT_SET_FONT_BOLD_ITALIC:
         case VTE_OSC_URXVT_VIEW_UP:
         case VTE_OSC_URXVT_VIEW_DOWN:
-#if _VTE_GTK != 3
-        case VTE_OSC_URXVT_EXTENSION:
-#endif
+//#if _VTE_GTK != 3
+        //case VTE_OSC_URXVT_EXTENSION:
+//#endif
         case VTE_OSC_YF_RQGWR:
         default:
                 break;
